Implement Lingering potion & Area Effect Cloud

IvanCraft623 opened the pull request:
## Introduction
Implements lingering potions and area effect cloud + custom effect support, also complements #5163


## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->
* The following new classes have been added:
  * `entity\effect\EffectContainer`
  * `pocketmine\entity\object\AreaEffectCloud`
  * `pocketmine\item\LingeringPotion`
* The following API methods have changed signatures
  * `SplashPotion->setLinger()` - now returns `self` instead of `void`

#### Rationale for `EffectContainer` addition
Unlike entities (Living) there are different types of things that can contain effects that are not necessarily applied. Such is the case of AreaEffectCloud.

Other stuff that may contain effects:
* Potion: https://github.com/pmmp/PocketMine-MP/blob/d5762d3f4487772ffb6b7c77af4cd659585f6520/src/item/Potion.php#L63
* Tipped arrows

## Backwards compatibility
<!-- Any possible backwards incompatible changes? How are they solved, or how can they be solved? -->
No BC-breaking changes.

## Tests
https://youtu.be/cRAjwEVbb0w
This has been playtested and and there was only one issue:
* On fire animation is shown when extinguishing fire (this is an already known bug on PM).

IvanCraft623 committed:
Implement lingering potion

IvanCraft623 committed:
Add LingeringPotion doc

IvanCraft623 committed:
Implement an EffectContainer

IvanCraft623 committed:
Cauldron integration

dktapps commented:
Nice, I'll review this tomorrow.

IvanCraft623 committed:
Constants usage! :D

IvanCraft623 committed:
Vanilla parity

IvanCraft623 commented:
Minecraft Bedrock implements AreaCloudEffect in a weird way, behaves differently than described in the wiki and in java.
Some examples:
* [ ] The number of times the effect is applied varies per effect (I still can't figure out the criteria).
* [x] Regardless of where the `SplashPotion` falls, the cloud will be positioned in the middle of the block.
* [x] `ReapplicationDelay` changes depending if the effect is Instantaneous.
* [x] `RadiusOnUse` is applied even the cloud has no effects.
* [x] The potion effect is stored in the `ListTag` effects, it does not make sense since it is aready storing the `PotionId`.
* [x] It is only updated every 10 ticks.

IvanCraft623 committed:
More vanilla parity!

IvanCraft623 committed:
Match vanilla properties

IvanCraft623 committed:
Use RADIUS_PER_TICK const

dktapps commented on src/entity/EntityFactory.php:
```diff
@@ -81,6 +82,14 @@ public function __construct(){
+				throw new SavedDataLoadingException("No such potion type");
+			}
+			return new AreaEffectCloud(Helper::parseLocation($nbt, $world), $potionType, $nbt);
+		}, ['AreaEffectCloud', 'minecraft:area_effect_cloud'], LegacyIds::AREA_EFFECT_CLOUD);
```
Are you sure these are correct? Don't make any assumptions - sometimes the legacy IDs in particular have seriously weird names.

IvanCraft623 replied:
Yeah, vanilla nbt tree:
![image](https://user-images.githubusercontent.com/57236932/190538649-aad8664f-790d-4e1a-8801-9b8de5b08661.png)

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,385 @@
+	/**
+	 * Recalculates the area entity's potion bubbles colour based on the active effects.
+	 */
+	protected function recalculateEffectColor() : void{
```
Why is this logic duplicated? The same already exists in `EffectManager`.

IvanCraft623 replied:
Not really, `EffectManager`'s ignores the color if the effect does not have bubbles and the default color is different.

dktapps replied:
That's one detail. It doesn't require you to duplicate the whole section of code.

IvanCraft623 replied:
I added `setDefaultBubbleColor` and `setEffectValidatorForBubbles` api methods, I would like to know if you find them good enough to be able to mark this conversation as resolved.

IvanCraft623 committed:
A lot of changes

- Move bubble color calculation to EffectContainer
- Implement API for effect validator for bubbles
- Implement API for default bubbles color
- Match more AreaEffectCloud properties with vanilla

IvanCraft623 committed:
shut

IvanCraft623 committed:
Shut!

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,384 @@
+		$this->networkPropertiesDirty = true;
+	}
+
+	public function getRadiusPerTick() : float{
```
This name is unclear and doesn't describe what the function actually returns. It should either be documented or renamed.

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,384 @@
+		$this->networkPropertiesDirty = true;
+	}
+
+	public function getWaiting() : int{
```
Same again here.

dktapps replied:
`getInitialApplicationAge()` would be a better name. This is actually supposed to be a "time until active" delay.

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,384 @@
+		$this->networkPropertiesDirty = true;
+	}
+
+	public function getDuration() : int{
```
and here

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,384 @@
+		$this->networkPropertiesDirty = true;
+	}
+
+	public function getDurationOnUse() : int{
```
and here

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,384 @@
+	/**
+	 * Sets the initial radius.
+	 *
+	 * @throws \RuntimeException if something attempted to set when it is already ticking
```
I don't like this. If the radius needs to be set in advance, it should be passed in the constructor.

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,384 @@
+		$spawnTime = $this->getWorld()->getTime() - $this->age;
+		$properties->setInt(EntityMetadataProperties::AREA_EFFECT_CLOUD_SPAWN_TIME, $spawnTime);
+		$properties->setInt(EntityMetadataProperties::AREA_EFFECT_CLOUD_DURATION, $this->duration);
+		$properties->setFloat(EntityMetadataProperties::AREA_EFFECT_CLOUD_RADIUS, $this->initialRadius);
```
Why isn't this using the current radius?

IvanCraft623 replied:
The client expects InitialRadius, as I mentioned in a comment the client calculates the radius on its own using SpawnTime, InitialRadius and some other stuff. If we send the current radius it will result in that on the client side the cloud will disappear much earlier than it should. See: https://github.com/df-mc/dragonfly/issues/665

IvanCraft623 replied:
See: https://github.com/df-mc/dragonfly/issues/665#issuecomment-1338609858

IvanCraft623 replied:
A hack has been implemented which consists of sending incorrect values in the entity properties `AREA_EFFECT_CLOUD_RADIUS_CHANGE_ON_PICKUP` and `AREA_EFFECT_CLOUD_RADIUS_PER_TICK`, to disable client-sided radius calculation.

dktapps reviewed (commented):
This PR is rather bloated by the EffectContainer changes so it's difficult for me to see what's going on. I've given this only a cursory review.

IvanCraft623 committed:
Merge branch 'next-major' into lingering_potion

IvanCraft623 committed:
Move initial radius to constructor

IvanCraft623 committed:
shut

IvanCraft623 committed:
Fix radius

IvanCraft623 committed:
Improve documentation

IvanCraft623 commented:
I have improved the documentation, because I am not a native speaker I would like if someone could verify that it has no grammatical errors and makes sense.

IvanCraft623 committed:
Merge branch 'next-major' into lingering_potion

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,416 @@
+	public const TAG_RADIUS_CHANGE_ON_PICKUP = "RadiusChangeOnPickup"; //TAG_Float
+	public const TAG_RADIUS_ON_USE = "RadiusOnUse"; //TAG_Float
+	public const TAG_RADIUS_PER_TICK = "RadiusPerTick"; //TAG_Float
+	public const TAG_EFFECTS = "mobEffects"; //TAG_List
```
These constants should not be public unless used outside the class. They are of no concern to plugin developers.

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,416 @@
+	public const DURATION_ON_USE = 0;
+
+	public const WAIT_TIME = 10;
+	public const REAPPLICATION_DELAY = 40;
```
These numbers are meaningless without units.

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,416 @@
+		return false;
+	}
+
+	public function getAge() : int{
```
Again, meaningless without units. Please document the function.

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,416 @@
+	}
+
+	/**
+	 * Returns the initial radius.
```
Radius in what unit?

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,416 @@
+					$this->setDuration($duration);
+				}
+			}
+			$this->setWaiting($this->age + self::WAIT_TIME);
```
I think you're using this incorrectly. This will prevent the effect cloud from applying effects for 30 seconds at a time, which defeats the purpose of having `reapplication delay` of 2 seconds.

According to the wiki: The time before deploying the field. The Radius is ignored, meaning that any specified effects is not applied and specified particles appear only at the center of the field, until Age hits this number.

IvanCraft623 replied:
I tested for this in vanilla, instant effects have a `reapplication Delay` of 0, but due to `wait Time` they are only applied every 10 ticks. So I noticed that `wait Time` is not only used during the first ticks of the entity's life. I also saw the java code and although the `wait Time` there is 5 ticks, `wait Time` also works as a delay for applying effects.

So...

`WaitTime`: Cloud Delay to apply effects, or update.
`ReapplicationDelay`: Delay per entity to reapply the effects, capped by `WaitTime`.

dktapps committed:
Rearrange and rename some properties

IvanCraft623 committed:
Mark as protected NBT tags

IvanCraft623 committed:
Document units

IvanCraft623 committed:
Avoid age underflow

IvanCraft623 committed:
Implement AreaEffectCloudApplyEvent

IvanCraft623 committed:
shut PHPStan!

dktapps commented on src/event/entity/AreaEffectCloudApplyEvent.php:
```diff
@@ -0,0 +1,69 @@
+	/**
+	 * Returns the effected entities.
+	 *
+	 * It is important to note that not every entity is guaranteed to be effected.
```
```suggestion
	 * It is important to note that not every entity is guaranteed to be affected.
```

dktapps replied:
I don't like this. It's not particularly useful without certainty that something is actually going to happen to all the entities as a result of the event.

dktapps commented on src/event/entity/AreaEffectCloudApplyEvent.php:
```diff
@@ -0,0 +1,69 @@
+	 * Returns the effected entities.
+	 *
+	 * It is important to note that not every entity is guaranteed to be effected.
+	 * The cloud may die during the application of its effects due to the depletion
```
```suggestion
	 * The cloud may despawn during the application of its effects due to the depletion
```

dktapps commented on src/event/entity/AreaEffectCloudApplyEvent.php:
```diff
@@ -0,0 +1,69 @@
+	 *
+	 * It is important to note that not every entity is guaranteed to be effected.
+	 * The cloud may die during the application of its effects due to the depletion
+	 * of {@link AreaEffectCloud::getDurationChangeOnUse()} or {@link AreaEffectCloud::getRadiusChangeOnUse()}.
```
```suggestion
	 * by {@link AreaEffectCloud::getDurationChangeOnUse()} or {@link AreaEffectCloud::getRadiusChangeOnUse()}.
```

IvanCraft623 committed:
Apply suggestions from dktapps

Co-authored-by: Dylan T. <odigiman@gmail.com>

IvanCraft623 committed:
Ensure that all entities returned by `AreaEffectCloudApplyEvent::getAffectedEntities()` will be affected.

IvanCraft623 committed:
Merge branch 'next-major' into lingering_potion

IvanCraft623 committed:
Hack to disable client side radius calculadtion

IvanCraft623 committed:
Merge branch 'next-major' into lingering_potion

IvanCraft623 committed:
Merge branch 'minor-next' into lingering_potion

IvanCraft623 committed:
Merge remote-tracking branch 'upstream/minor-next' into lingering_potion

IvanCraft623 committed:
Merge branch 'minor-next' into lingering_potion

IvanCraft623 committed:
Use of native enum case accessor

dktapps commented on src/entity/projectile/SplashPotion.php:
```diff
@@ -164,10 +177,13 @@ public function willLinger() : bool{
+	 * @return $this
 	 */
-	public function setLinger(bool $value = true) : void{
+	public function setLinger(bool $value = true) : self{
```
This is a BC break for inheritors of SplashPotion

dktapps commented on src/entity/effect/EffectManager.php:
```diff
@@ -91,60 +53,29 @@ public function remove(Effect $effectType) : void{
-		$oldEffect = null;
-		$cancelled = false;
-
+	public function add(EffectInstance $effect, bool $force = false) : bool{
```
This parameter addition is also BC breaking for inheritors

IvanCraft623 committed:
Merge branch 'minor-next' into lingering_potion

IvanCraft623 committed:
Make `SplashPotion::setLinger()` not fluent :(

Always BC...

IvanCraft623 committed:
Remove `EffectManager::add()` `$force` param

:(

IvanCraft623 committed:
Ugh remove $force param from EffectContainer too

this was useful to skip unnecesary checks :(

IvanCraft623 committed:
...

dktapps commented on src/item/LingeringPotion.php:
```diff
@@ -0,0 +1,69 @@
+		parent::__construct($identifier, $name);
+	}
+
+	protected function describeType(RuntimeDataReader|RuntimeDataWriter $w) : void{
```
This should be `RuntimeDataDescriber`, we're not in PM5 alpha anymore

See https://github.com/pmmp/PocketMine-MP/blob/49da50659f08472ddc41b6925098006ba95de3f5/src/item/SplashPotion.php#L36-L38

dktapps commented on src/item/LingeringPotion.php:
```diff
@@ -0,0 +1,69 @@
+	private PotionType $potionType;
+
+	public function __construct(ItemIdentifier $identifier, string $name){
+		$this->potionType = PotionType::WATER;
```
This can be a property default directly, it's a native enum now

dktapps commented on src/item/StringToItemParser.php:
```diff
@@ -1202,6 +1202,7 @@ private static function registerDynamicItems(self $result) : void{
 
 			$result->register($prefix("potion"), fn() => Items::POTION()->setType($potionType));
 			$result->register($prefix("splash_potion"), fn() => Items::SPLASH_POTION()->setType($potionType));
+			$result->register($prefix("lingering_potion"), fn() => Items::SPLASH_POTION()->setType($potionType));
```
```suggestion
			$result->register($prefix("lingering_potion"), fn() => Items::LINGERING_POTION()->setType($potionType));
```

IvanCraft623 committed:
Modernize code on item\LingeringPotion

IvanCraft623 committed:
StringToItemParser: SPLASH_POTION => LINGERING_POTION

my bad :/

Co-authored-by: Dylan T. <odigiman@gmail.com>

IvanCraft623 committed:
Merge branch 'lingering_potion' of https://github.com/IvanCraft623/PocketMine-MP into lingering_potion

dktapps commented on src/entity/effect/EffectContainer.php:
```diff
@@ -0,0 +1,239 @@
+
+	/**
+	 * In the following cases it will return true:
+	 * If a weaker effect of the same type is already applied.
+	 * If a weaker or equal-strength effect is already applied but has a shorter duration.
```
```suggestion
	 * - if the effect type is not already applied
	 * - if an existing effect of the same type can be replaced (due to shorter duration or lower level)
```

dktapps commented on src/entity/effect/EffectContainer.php:
```diff
@@ -0,0 +1,239 @@
+	 * If a weaker effect of the same type is already applied.
+	 * If a weaker or equal-strength effect is already applied but has a shorter duration.
+	 *
+	 * @return bool whether the effect can been applied.
```
This seems redundant

dktapps commented on src/entity/effect/EffectContainer.php:
```diff
@@ -0,0 +1,239 @@
+	/**
+	 * @phpstan-param \Closure(EffectInstance) : bool $effectValidator
+	 */
+	public function setEffectValidatorForBubbles(\Closure $effectValidator) : void{
```
I think this would be better named "effect filter" rather than "effect validator".

dktapps commented on src/entity/effect/EffectContainer.php:
```diff
@@ -0,0 +1,239 @@
+
+	public function getDefaultBubbleColor() : Color{
+		return $this->defaultBubbleColor;
+	}
```
I don't like this. It's only required by AreaEffectCloud.

Really we should check if `getBubbleColor()` is null and then fill it with a default on AreaEffectCloud (or, for the sake of BC, check if the RGBA colour code is 0).

IvanCraft623 replied:
I hate having a RGBA(0) and having to do a check on network layer to display the correct color, it would consufuse plugin-devs when doing `getBubbleColor()` on water `AreaEffectCloud`s. but is the only way for BC :c

IvanCraft623 replied:
I mean I really prefer the current solution, EffectManager/Container in the first place shouldn't be setting a RGBA(0) when there are no effects

dktapps replied:
> EffectManager/Container in the first place shouldn't be setting a RGBA(0) when there are no effects

Agreed, but I don't think adding more conditionally useless stuff to its API is a good solution. We're gonna have to BC break at some point to fix this anyway.

SplashPotion already suffers from this issue anyway as seen here: https://github.com/pmmp/PocketMine-MP/blob/012894764c1e87569a893014db3ad8e8223f0817/src/entity/projectile/SplashPotion.php#L83-L95

dktapps replied:
To be honest, looking at SplashPotion I'm not really sure why we want to do this EffectContainer refactor anyway. I'm doubtful it makes sense to allow mutating the effects given the increase in complexity it causes.

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,462 @@
+		parent::initEntity($nbt);
+
+		$this->effectContainer = new EffectContainer();
+		$this->effectContainer->setDefaultBubbleColor(PotionSplashParticle::DEFAULT_COLOR());
```
As above, don't like this. Default colour of bubbles shouldn't be the EffectContainer's responsibility, since it's only going to be used by AreaEffectCloud.

IvanCraft623 committed:
Updade EffectContainer::canAdd() documentation

Co-authored-by: Dylan T. <dktapps@pmmp.io>

IvanCraft623 committed:
Remove @return

dktapps commented on src/item/LingeringPotion.php:
```diff
@@ -0,0 +1,63 @@
+
+	protected function createEntity(Location $location, Player $thrower) : Throwable{
+		$projectile = new SplashPotionEntity($location, $thrower, $this->potionType);
+		$projectile->setLinger();
```
I wonder if we could just add a `bool $linger` parameter to the constructor of `SplashPotion` for this. The two classes are identical apart from this function.

IvanCraft623 replied:
Adding it on constructor since the class is not final would be a BC break for inheritors. but we could try a `setLinger()` method.

dktapps replied:
It's not BC breaking unless it's required.

dktapps replied:
`setLinger()` isn't an option because we already have a type ID constant for lingering potion. It would be a BC break to remove it, and confusing to plugin devs not to remove it.

IvanCraft623 committed:
Rename EffectContainer::setEffectValidatorForBubbles() method

and add documentation

IvanCraft623 committed:
Introduce $linger param on SplashPotion

this removes code duplication between LingerPotion and SpashPotion

dktapps commented on src/item/SplashPotion.php:
```diff
@@ -52,10 +62,19 @@ public function getMaxStackSize() : int{
 	}
+
+	/**
+	 * Returns whether this splash potion will create an area-effect cloud when it lands on it's projectile form.
```
```suggestion
	 * Returns whether this splash potion will create an area-effect cloud on impact.
```

dktapps commented on src/item/VanillaItems.php:
```diff
@@ -514,6 +515,7 @@ protected static function setup() : void{
 		self::register("lapis_lazuli", fn(IID $id) => new Item($id, "Lapis Lazuli"));
 		self::register("lava_bucket", fn(IID $id) => new LiquidBucket($id, "Lava Bucket", Blocks::LAVA()));
 		self::register("leather", fn(IID $id) => new Item($id, "Leather"));
+		self::register("lingering_potion", fn(IID $id) => new SplashPotion($id, "Lingering Potion", [], true));
```
```suggestion
		self::register("lingering_potion", fn(IID $id) => new SplashPotion($id, "Lingering Potion", linger: true));
```

dktapps commented on src/entity/effect/EffectContainer.php:
```diff
@@ -0,0 +1,239 @@
+
+	protected Color $defaultBubbleColor;
+
+	protected ?Color $bubbleColor = null;
```
This is also BC breaking for inheritors (the original field wasn't nullable)

dktapps commented on src/entity/effect/EffectContainer.php:
```diff
@@ -0,0 +1,239 @@
+	 *
+	 * @phpstan-var \Closure(EffectInstance) : bool
+	 */
+	protected \Closure $effectValidatorForBubbles;
```
This should be renamed to use the filter terminology

dktapps commented on src/entity/projectile/SplashPotion.php:
```diff
@@ -126,10 +127,22 @@ protected function onHit(ProjectileHitEvent $event) : void{
+			if(($owner = $this->getOwningEntity()) !== null && !$owner->isClosed()){
+				$entity->setOwningEntity($owner);
+			}
+			foreach($this->getPotionEffects() as $effect){
+				$entity->getEffects()->add($effect);
+				if($effect->getType() instanceof InstantEffect){
+					$entity->setReapplicationDelay(0);
+				}
+			}
```
I don't like this.

Currently we have an AEC entity that takes a PotionType in the constructor, but the potion type is useless because the entity doesn't use it. No effects are taken from the potion type. Good way to troll plugin devs trying to spawn AECs with a specific potion type.

dktapps commented on src/entity/object/AreaEffectCloud.php:
```diff
@@ -0,0 +1,462 @@
+	 * Returns the effects the area effect cloud provides.
+	 *
+	 * Used to get COPIES to avoid accidentally modifying the same effect instance
+	 * already applied to another entity.
```
No one is gonna read this. They'll just use `getEffects()` because that's what they're familiar with.

IvanCraft623 committed:
Update SplashPotion::willLinger() doc

Co-authored-by: Dylan T. <dktapps@pmmp.io>

IvanCraft623 committed:
Skip $enchantmentTags param on SplashPotion construction :P

Co-authored-by: Dylan T. <dktapps@pmmp.io>

IvanCraft623 committed:
Get rid of EffectContainer::defaultBubbleColor

IvanCraft623 committed:
Add missing filter -> validator terminology changes

dktapps committed:
Merge remote-tracking branch 'origin/minor-next' into lingering_potion

dktapps committed:
Cleanup, terminology improvements, make AreaEffectCloud constructor behaviour less surprising

dktapps committed:
Remove crap that has no business being sent over the network

dktapps committed:
we're not syncing this over the network anymore

dktapps committed:
Remove useless property

dktapps committed:
oops

dktapps committed:
Remove internal function

dktapps committed:
Fucking CS

dktapps commented:
This is stalled pending some verification about the usage of the NBT save data. It's not clear how `PotionId` interacts with `mobEffects`.

kostamax27 commented:
cyclic references in:

https://github.com/IvanCraft623/PocketMine-MP/blob/d0647099a69e136ed4854cbb4bcb554789853e95/src/entity/effect/EffectContainer.php#L70

also in [AreaEffectCloud](https://github.com/IvanCraft623/PocketMine-MP/blob/d0647099a69e136ed4854cbb4bcb554789853e95/src/entity/object/AreaEffectCloud.php) is missing `unset(effectContainer)` in `destroyCycles`

dktapps committed:
Merge branch 'minor-next' into lingering_potion

dktapps committed:
tidy

dktapps committed:
fix

dktapps committed:
fix broken things
